File: server/models/vaultModel.js
const mongoose = require('mongoose');

const vaultSchema = new mongoose.Schema(
  {
    userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    name: { type: String, required: true, trim: true },
    originalName: { type: String, required: true },
    mimeType: {
      type: String,
      required: true,
      enum: [
        'application/pdf',
        'text/csv',
        'application/vnd.ms-excel',                                          // .xls
        'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet', // .xlsx
      ],
    },
    size: { type: Number, required: true }, // bytes
    data: { type: String, required: true }, // base64 encoded
    tags: [{ type: String, trim: true }],
    description: { type: String, trim: true, default: '' },

    // Reserved for RAG pipeline (handled by Python server)
    isProcessedForRAG: { type: Boolean, default: false },

    // Password-protected PDF handling
    // cron sets passwordProtected=true when it fails to parse due to encryption
    // user supplies password via unlock endpoint; cron uses it and clears it after embedding
    passwordProtected: { type: Boolean, default: false },
    pdfPassword:       { type: String,  default: '' },
  },
  { timestamps: true }
);

// Enforce 16MB limit at schema level
vaultSchema.pre('save', function (next) {
  const MAX_SIZE = 16 * 1024 * 1024; // 16MB
  if (this.size > MAX_SIZE) {
    return next(new Error('File size exceeds 16MB limit'));
  }
  next();
});

module.exports = mongoose.model('Vault', vaultSchema);

File: chatServer/app/models/vaultModel.py
"""
VaultModel — mirrors the Node.js Vault mongoose schema.
Python RAG pipeline reads from this collection (read-only).
Node.js owns writes; Python only updates `isProcessedForRAG`.

Supported mimeTypes (must match Node.js vaultModel allowlist):
  application/pdf
  text/csv
  application/vnd.openxmlformats-officedocument.spreadsheetml.sheet  (XLSX)
"""
from pydantic import BaseModel, Field, ConfigDict, field_validator
from typing import Optional, List, Literal
from datetime import datetime
from bson import ObjectId

from app.models.base import PyObjectId

# Matches the Node.js mongoose enum exactly
VaultMimeType = Literal[
    "application/pdf",
    "text/csv",
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
]


class VaultModel(BaseModel):
    """
    Read model for the existing Vault collection owned by Node.js.
    Fields match the mongoose schema 1-to-1.
    """

    id: Optional[str] = Field(None, alias="_id")
    userId: str
    name: str                           # display name
    originalName: str                   # original filename
    mimeType: VaultMimeType = "application/pdf"
    size: int                           # bytes
    data: str                           # base64-encoded file content
    tags: List[str] = Field(default_factory=list)
    description: str = ""

    # RAG pipeline flag — Python sets this True after embedding
    isProcessedForRAG: bool = False

    createdAt: Optional[datetime] = None
    updatedAt: Optional[datetime] = None

    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True,
        json_encoders={ObjectId: str},
    )

    @field_validator("userId", "id", mode="before")
    @classmethod
    def coerce_objectid_to_str(cls, v):
        return str(v) if isinstance(v, ObjectId) else v


class VaultListItem(BaseModel):
    """Lightweight projection for listing vaults — avoids loading base64 data."""

    id: Optional[str] = Field(None, alias="_id")
    userId: str
    name: str
    originalName: str
    mimeType: VaultMimeType
    size: int
    tags: List[str] = Field(default_factory=list)
    description: str = ""
    isProcessedForRAG: bool = False
    createdAt: Optional[datetime] = None
    updatedAt: Optional[datetime] = None

    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True,
        json_encoders={ObjectId: str},
    )

    @field_validator("userId", "id", mode="before")
    @classmethod
    def coerce_objectid_to_str(cls, v):
        return str(v) if isinstance(v, ObjectId) else v